hedera-node/hedera-app/src/main/java/com/hedera/node/app/blocks/impl/GrpcBlockItemWriter.java

```diff
@@ -0,0 +1,214 @@
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Implements the bidirectional streaming RPC for the publishBlockStream rpc in BlockStreamService
+ * See <a href="https://grpc.io/docs/languages/java/basics/">gRPC Basics</a>
+ */
+public class GrpcBlockItemWriter implements BlockItemWriter {
+
+    private static final Logger logger = LogManager.getLogger(GrpcBlockItemWriter.class);
+    private static final String INVALID_MESSAGE = "Invalid protocol buffer converting %s from PBJ to protoc for %s";
+
+    private final BlockStreamServiceGrpc.BlockStreamServiceStub asyncStub;
+    private final ManagedChannel channel;
+    private StreamObserver<PublishStreamRequest> requestObserver;
+
+    private long blockNumber;
+
+    /** The state of this writer */
+    private State state = State.UNINITIALIZED;
+
+    /**
+     * The current state of the gRPC writer.
+     */
+    public enum State {
+        /**
+         * The gRPC client still not initialized
+         */
+        UNINITIALIZED,
+        /**
+         * The gRPC client is currently open
+         */
+        OPEN,
+        /**
+         * The gRPC client is already closed
+         */
+        CLOSED
+    }
+
+    /**
+     * @param blockStreamConfig the block stream configuration
+     */
+    public GrpcBlockItemWriter(@NonNull final BlockStreamConfig blockStreamConfig) {
+        requireNonNull(blockStreamConfig, "The supplied argument 'blockStreamConfig' cannot be null!");
+
+        channel = ManagedChannelBuilder.forAddress(blockStreamConfig.address(), blockStreamConfig.port())
+                .usePlaintext()
+                .build();
+        asyncStub = BlockStreamServiceGrpc.newStub(channel);
+    }
+
+    @Override
+    public void openBlock(long blockNumber) {
+        if (state != State.UNINITIALIZED) throw new IllegalStateException("GrpcBlockItemWriter initialized twice");
+
+        if (blockNumber < 0) throw new IllegalArgumentException("Block number must be non-negative");
+        this.blockNumber = blockNumber;
+        requestObserver = asyncStub.publishBlockStream(new StreamObserver<>() {
+            @Override
+            public void onNext(PublishStreamResponse streamResponse) {
+                // close the stream if error stream response occurs - source handshake design
+                // https://github.com/hashgraph/hedera-block-node/issues/120
+                /*if (streamResponse.getStatus().getStatus() == STREAM_ITEMS_UNKNOWN) {
+                 */
+                /*, STREAM_ITEMS_SUCCESS, STREAM_ITEMS_BEHIND*/
+                /*
+                    onNext(PublishStreamResponse.newBuilder()
+                            .setStatus(EndOfStream.newBuilder()
+                                    .setStatus(STREAM_ITEMS_UNKNOWN)
+                                    .build())
+                            .build());
+
+                    closeBlock();
+                }*/
```

Comment:
Todo later?

Reply:
Those were for testing purposes only, currently I removed them but we will need to implement handlings of different scenarios based on the returned `EndOfStream` from block nodes